### PR TITLE
Allow the window to shrink in vim.exe

### PR DESCRIPTION
**Pull request by ghost**
In vim.exe, when the window is shrunk manually, the event does not occur, so the edit area is not shrunk.
So, at the timing of redrawing the edit screen, I tried to reduce the edit area.
I confirmed the operation with Windows Vista / 7/10 (1809) / 10 (1903).

### CODE REVIEW

File: src/os_win32.c [hunk 1]
```diff
@@ -7893,3 +7893,29 @@ is_conpty_stable(void)
 {
     return conpty_stable;
 }
+
+#ifndef FEAT_GUI_W32
+
+void redraw_ncwindow(void)
```
Comment (k-takata) [inline src/os_win32.c]: What does `nc` mean?

Comment (ghost) [reply src/os_win32.c]: Non-client like WM_NCPAINT etc...

Comment (k-takata) [reply src/os_win32.c]: I don't think that this function redraws non-client window.
Just resize the console buffer to fit with the console window, doesn't it?
`resize_console_buf` or something might be better?

Comment (ghost) [reply src/os_win32.c]: I have written the purpose as it is. I should have written the behavior.
I'll fix it now.

**Comment by ghost** [comment]
I need to recognize that I am careless.

**Comment by GreenCourt** [comment]
Can this PR fix https://github.com/vim/vim/issues/3611 ?

**Comment by k-takata** [comment]
I tried this PR on Win10.
When I shrink the size of the window, Vim's display follows the size. It is good.
However, the height of the console buffer still keeps the old height, and when I type Ctrl-L, the buffer height fits to the window height. It would be nice if the console buffer size automatically fits to the window size.

**Comment by ghost** [comment]
I'm sorry, I did not know #3611.
After all, it is a fix of it.

The buffer size fit is under challenge.

**Comment by k-takata** [comment]
I created a different patch for this: https://gist.github.com/k-takata/feb187996d023b243eaf4eb0a01e6bc8
I tested this only on Windows 10. I don't know whether this works on Win7, etc.

**Comment by ghost** [comment]
Thank you very much. I tried it.
It works normally: Win10 (19H1, 20H1)
It does not work: Win10 (1809), Win7, Vista

I mixed it in PR. For OSs that do not automatically fit, CTRL+L is valid. Do you mind if this is done?
I would like a joint name.

**Comment by k-takata** [comment]
> It does not work: Win10 (1809), Win7, Vista

Oh, really? It worked on Win10 1709 and 1809 in my side.

On 1709, `WriteConsoleOutputCharacterW()` sometimes returns -1 into `cchwritten`. It is undocumented, but I had to add the check to avoid crash. This doesn't happen on 1809.
I haven't checked if the ANSI API also returns -1, but it might be better to add the check.

```diff
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -6331,7 +6339,7 @@ write_chars(
 	     * character was written, otherwise we get stuck. */
 	    if (WriteConsoleOutputCharacter(g_hConOut, (LPCSTR)pchBuf, cbToWrite,
 			coord, &written) == 0
-		    || written == 0)
+		    || written == 0 || cchwritten == (DWORD)-1)
 		written = 1;
 	}
 	else
```

**Comment by ghost** [comment]
Ah. Successful if the "Wrap text output on resize" checkbox is enabled.
Because it was Insider build, it was not.
I was disabled, I did not notice. I'm sorry.

If enabled, it works on Win10 (1809). Win7 and Vista do not have checkboxes.

**Comment by ghost** [comment]
There seems to be no problem in operation. I must be careful about checkboxes.
Is this patch still changed?

**Comment by k-takata** [comment]
I think it's okay.

**Pull request closed without merging**